### Kaggle ML and Data Science Survey Analysis
#### Data 512, Final Project Plan - Zicong Liang

### Project Motivation
This project an analysis for a survey about Machine Learning and Data Science. Recently, lots of people are talking about machine learning and Data Science. In addition, more and more companies hire data science talents and invest in data science in order to make their business more data-driven. According to Wikipedia, Data science is also known as data-driven science, is an interdisciplinary field combined from mathematics, statistics and computer science to extract insights from data. On the one hand, I remember Oliver commented about nobody's really done a proper survey related to Data Science. I think it is a great opportunity to do some research in this field of study because I find out Kaggle releases a survey responses about Data Science and Machine Learning. On the other hand, there are a few reasons drive me to know more about data science and machine learning field of study.
1.	I am one of the graduate students in the master of science data science program at University of Washington.
2.	I'd like to get a job in data science filed after graduating from this program.
3.	I'd like to know how those people related to this field think about data science and machine learning, then analyses their response to draw some conclusions that I am looking for. 

As one of the beginner in Data Science, it is better to know more about what kind of skills we need to well-equipped in order to be ready to step into this field. That's why I am planning to do this analysis.
Throughout this project, we are going to conduct some analyses for Data Science and Machine Learning in various aspects, such as gender, age, programming skills, algorithms, education degree and salary and so on. It helps us gain new insights of Data Science that different from what we have learnt from school. In addition, the analysis of this project can be a reference to those who people for pursuing a career in Data Science, not only for switching job but also looking for their first job in this field.

### Project Dataset
The data I am going to use for this project is from Kaggle. It is one of the most popular datasets in Kaggle recently. To be more specifically, the data is about a survey which conducted by Kaggle for an industry-wide to establish a comprehensive view of the state of Data Science and Machine Learning. This survey received more than 16000 responses. According to this data, we would be able to learn a ton about who is working in the Data Science Field, what’s happening at the cutting edge of machine learning across industries, and how new data scientists can best break into the field.

Here is the link to the main page of [Kaggle ML and Data Science Survey, 2017](https://www.kaggle.com/kaggle/kaggle-survey-2017).

The dataset consists of 5 files:
1. **schema.csv**: a CSV file with survey schema. This schema includes the questions that correspond to each column name in both the **multipleChoiceResponses.csv** and **freeformResponses.csv**.
2. **multipleChoiceResponses.csv**: Respondents' answers to multiple choice and ranking questions.
3. **freeformResponses.csv**: Respondents' freeform answers to Kaggle's survey questions.
4. **conversionRates.csv**: Currency conversion rates to USD (accessed from the R package "quantmod" on September 14, 2017)
5. **RespondentTypeREADME.txt**: This is a schema for decoding the responses in the "Asked" column of the **schema.csv** file.

Click [here](https://www.kaggle.com/kaggle/kaggle-survey-2017/data) to the data page.

Base on the description above, I am going to use two files from the data.
1. **multipleChoiceResponses.csv**
2. **conversionRates.csv**
The **conversionRates.csv** data is pretty straightforward, so let's look at the **multipleChoiceResponses.csv** before the start any analysis.

In [31]:
import pandas as pd
my_data = pd.read_csv("multipleChoiceResponses.csv", encoding='ISO-8859-1', delimiter=',', low_memory=False)
my_data.head()

GenderSelect        Country   Age  \
0  Non-binary, genderqueer, or gender non-conforming            NaN   NaN   
1                                             Female  United States  30.0   
2                                               Male         Canada  28.0   
3                                               Male  United States  56.0   
4                                               Male         Taiwan  38.0   

                                    EmploymentStatus StudentStatus  \
0                                 Employed full-time           NaN   
1                 Not employed, but looking for work           NaN   
2                 Not employed, but looking for work           NaN   
3  Independent contractor, freelancer, or self-em...           NaN   
4                                 Employed full-time           NaN   

  LearningDataScience CodeWriter CareerSwitcher  \
0                 NaN        Yes            NaN   
1                 NaN        NaN            NaN   
2                 NaN        NaN            NaN   
3                 NaN        Yes            NaN   
4                 NaN        Yes            NaN   

              CurrentJobTitleSelect TitleFit              ...                \
0             DBA/Database Engineer     Fine              ...                 
1                               NaN      NaN              ...                 
2                               NaN      NaN              ...                 
3  Operations Research Practitioner   Poorly              ...                 
4                Computer Scientist     Fine              ...                 

  JobFactorExperienceLevel JobFactorDepartment  JobFactorTitle  \
0                      NaN                 NaN             NaN   
1                      NaN                 NaN             NaN   
2           Very Important      Very Important  Very Important   
3                      NaN                 NaN             NaN   
4                      NaN                 NaN             NaN   

  JobFactorCompanyFunding JobFactorImpact JobFactorRemote JobFactorIndustry  \
0                     NaN             NaN             NaN               NaN   
1                     NaN             NaN             NaN               NaN   
2          Very Important  Very Important  Very Important    Very Important   
3                     NaN             NaN             NaN               NaN   
4                     NaN             NaN             NaN               NaN   

  JobFactorLeaderReputation JobFactorDiversity JobFactorPublishingOpportunity  
0                       NaN                NaN                            NaN  
1        Somewhat important                NaN                            NaN  
2            Very Important     Very Important                 Very Important  
3                       NaN                NaN                            NaN  
4                       NaN                NaN                            NaN  

[5 rows x 228 columns]

In [32]:
my_data.shape

(16716, 228)

We can see the first 5 rows from the above data (**multipleChoiceResponses.csv**). There are lots of variables that contain NaN values. This file contains 16716 observations and 228 variables. During the project, I will focus on the subset of this file, keeping those varibles that necessary for my analysis and dropping those NaN values.

**Note: Since there are 228 columns in the *multipleChoiceResponses.csv*, I might need further exploration of this file in order to determine how many variables I will need throughout this project. As long as I have my decision while answering the questions and hypotheses below, I would follow what I have done for the previous assignments to make a final dataset that include all the variables I need for this project as well as the descriptions.**

**Note: Since the data is from an online survey responses, I think bias would be possibly exist in the data, I will analysis the findings from data as well as combing facts to draw the my conclusions.**


#### License
Released UnderDatabase: [Open Database](http://opendatacommons.org/licenses/odbl/1.0/), Contents: © Original Authors

Here is some information about the Open Database License on Wikipedia. Click [here](https://en.wikipedia.org/wiki/Open_Database_License) to see more about it.

**Note: It seems there is some issue with the site, hopefully it will be fixed soon. (Mark at 4:47PM November 9th 2017)**

For more information about Kaggle please see [Terms](https://www.kaggle.com/terms) and [Privacy](https://www.kaggle.com/about/privacy)

### Research Question & Hypothesis

#### Research Question
I did some research that most people are using Python and R in the Data Science field, some use MATLAB and SAS. I'd 
As a student in the data science program, most of our courses suggest us to use Python as our main programming language instead of R. As I mention above, I'd like to become a data scientist after I graduate from the program, it's good to know what kind of companies that data scientists would most interested in working for and the salary of course. Moreover, I have listed some of the questions that raise in mind I want to know about from this research.

Q1: What's the gender diversity among this survey? (What's the difference between male and female in this field?)

Q2: As a student, really want to know much time should I spend studying in Data Science every week? And how important my MS Data Science Degree would help me success in this filed?

Q3: What is dominant programming language that people would suggestion to use or learn as in Data Science and Machine Learning, Python or R?

Q4: What are 10 most popular algorithms that people think are important to know as in the Data Science filed?

Q5: What are the challenges as working in Data Science filed that most people take for?

Q6: What's the mean/median salary as a data scientist? Moreover, what are the important factors that people consider would consider affecting the salary?

More questions might be added later

**Note: During the process of the project, I think there will be more interesting findings from the data (since there are lots of variables in the data file, I will explore more about the context and try to raise more related questions throughout the process in order to expand my analysis. **

#### Hypothesis
The hypothesis that first come to my mind is about gender imbalance.

*Hypothesis: There is a gender gap between male and female in Data Science field.*
Another hypothesis is about dominant programming language. 

*Hypothesis: Python is the most popular tool for data scientist compare to the other programming language.*

Moreover, I'd like to raise a hypothesis about the relationship between some important factors (e.g. skills, degree, experience) and salary. However, to address this hypothesis properly, I will need to finalize the variables selection.

*Hypothesis: There is relationship between some factors (e.g. skills, degree, experience) and salary.*

**Note: the project will focus on raising questions and try to answer those questions from the analysis more emphasis on using visulizations.**

### Technical Approach & Method

During this project, I am going to Using python and the following packages to acheive my analysis
1. **Pandas**
2. **Matplotlib**
3. **Seaborn**
4. **Scikit Learn**

The purpose of this project is to analyze the survey response and draw some conclusion base on the analysis. Therefore, I will use the Python Package Pandas to manipulate the data such as grouping, filtering and aggregating the data. Moreover, I will also use the Matplotlib and Seaborn libraries together to generate some visualizations from the data for the questions I raise above. As in human-centered Data Science, providing appropriate visualizations corresponding to the questions should be better for interpretation as well as easier for others to understand the questions and results. In the end, as time permits, I might also consider developing a regression model to determine what factors would impact the salary the most by using the Sklearn package (e.g. education degree, programming skills, machine learning algorithms, gender, experience and location). The final approach might be change a bit while working on the project (I might need more packages later if it is necessary). However, the overall analysis should be supported by visualizations as well as explanation to address the research questions and hypotheses.

To be more specifically, I will follow what I have learnt from the course HCDE 511 (human centered design & engineering) to encode the data properly for all the visualizations. Since there are lots of comparisons within each visualization, I think using bar charts (both vertical and horizontal) and pie charts should be the best choice to meet the requirement. Moreover, hue is also helpful to identify different categories in the visualizations.

### Any unknowns or dependencies

As I mention above, the data file (*multipleChoiceResponses.csv*) I am going to use  has 228 variables. I need further exploration about the content to determine how many variables are needed throughout this project. At the end, I will make a final data set that contains some of the original variable and might have some new generated variables (such as counting, mean, sum) which depend on the other variables. In addition, I will have a full description about the final data set. The analysis will base on the final dataset.